Unfinished tasks

FrostyX opened the pull request:
This PR adds a script that prints how many unfinished tasks are remaining in wiki file.

Example usage:

    [jkadlcik@chromie utils]$ ./vwunfinished.py --today --filetype md --count-sublists
    9

It also provides a convenient `vimwiki_unfinished_tasks` function that can be imported and used from some other python code.

The reason for this is ... status bars! I want to show how many remaining tasks for today I have, in my status bar.

brennen commented on vwunfinished.py:
```diff
@@ -0,0 +1,126 @@
+                             wiki_path=None, diary_dir=None, filetype=None):
+    """
+    Return a count of unfinished tasks in specified vimwiki file
+    Consider this function to be an API. It's attributes and return type will remain backwards compatible.
```
```suggestion
    Consider this function to be an API. Its attributes and return type will remain backwards compatible.
```

brennen commented on vwunfinished.py:
```diff
@@ -0,0 +1,126 @@
+            return os.path.join(*[os.path.expanduser(self.wiki_path), self.diary_dir,
+                                  ".".join([self.date, self.filetype])])
+
+        raise ValueError("I have not enough information to determine a file")
```
```suggestion
        raise ValueError("I do not have enough information to determine a file")
```

brennen commented on vwunfinished.py:
```diff
@@ -0,0 +1,126 @@
+        return tasks
+
+    def count_unfinished_tasks(self):
+        return len(self.unfinished_tasks)
```
Another suggestion for future work:  Since you have these anyway, it might be nice to provide an option to print out the full task list.

FrostyX commented on vwunfinished.py:
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/python
```
Fixed

FrostyX commented:
Hello @brennen,
thank you very much for the review.

I've committed the inline changes, that you suggested. I would like to squash the changes to the original commits before merging but keeping it in separate commits to make to review easier.

> Interface: Took me a minute to figure out the usage pattern, but I did finally get useful output:

Sorry for that, I wanted to properly describe it in a blog post or `README.md`, but didn't have enough time to do so (yet). Unfortunately, that cost your time, ... my apologies. Do you find the UI confusing, should I rework it? Or it was just about having to go through the argparser options?

> I almost wonder if the --count-sublists behavior should be the default, with an --ignore-sublists or similar? Maybe I have more nested tasklists than others.

That's probably a good idea. Personally, I always want to `--count-sublists`, so having it by default and rather occasionally specify `--ignore-sublists` would be more convenient for me. 

> Indentation: Actually, on looking further, it seems like --count-sublists is required if any tasks are indented at all, not just nested. For example, in my tasks.wiki, I have this list, and the top-level items don't get picked up:

I didn't realize, that someone might have indented-while-not-nested lists. I am not sure how to implement it without dramatically complicating the `unfinished_tasks` function. What do you think about adding a parameter saying, how much indentation should be stripped from the beginning of each line?

> All pages: As a general comment, I was sort of expecting this to search the entire wiki, rather than a page at a time. That could be scripted, of course, but it would be a useful addition in future.

This wouldn't be useful for my use case at all since I have gazillions of unfinished tasks spread through the entire wiki. However, I can see, that it might be useful for someone. Would you mind leaving it for separate PR? It should be a fairly straightforward implementation (not requiring `path` or `date` and in such case using `wiki_path` and go through all files).

> Testing: I tried the instructions in the README and some variations, and always wound up with something like:

Will installing `mock` package for python help? I am not sure what linux distribution you are using, on Fedora I would do `dnf install python3-mock` or in the worst possible scenario use `pip3 install mock`.

brennen commented:
> What do you think about adding a parameter saying, how much indentation should be stripped from the beginning of each line?

That seems like a very reasonable approach.  I think for my use case I'm always going to want to count all tasks, but I can see the use of limiting it...

> Will installing mock package for python help? I am not sure what linux distribution you are using, on Fedora I would do dnf install python3-mock or in the worst possible scenario use pip3 install mock.

Yep, sure enough.  Apologies I didn't specify what environment I was using in the review.

> This wouldn't be useful for my use case at all since I have gazillions of unfinished tasks spread through the entire wiki. However, I can see, that it might be useful for someone. Would you mind leaving it for separate PR?

Yeah, definitely seems out of scope for this PR.

FrostyX commented:
@brennen,
I've switched from `--count-sublists` to `--ignore-sublists` (which is by default `False`), and also implemented a `--indentation-level` for the use-case when someone has indented even top-level tasks and wishes to not count sublists.

Therefore, if I am not mistaken, everything we discussed that should be a part of this PR is now finished.

brennen commented:
@FrostyX right on - I don't have access to merge here, but it seems good to me if anyone with a commit bit feels like taking my word for it. :)

Nudin commented:
I merged it. If anyone wants to do further improvements just create a new PR.
